src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,264 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import ring_theory.dedekind_domain.adic_valuation
+
+/-!
+# The finite adèle ring of a Dedekind domain
+We define the ring of finite adèles of a Dedekind domain `R`.
+
+## Main definitions
+- `prod_completions_integers` : product of `adic_completion_integers`, where `v` runs over all
+   maximal ideals of `R`.
+- `prod_completions` : the product of `adic_completion`, where `v` runs over all maximal ideals
+  of `R`.
+- `finite_adele_ring` : The finite adèle ring of `R`, defined as the restricted product
+  `Π'_v K_v`.
+
+## Implementation notes
+We are only interested on Dedekind domains of Krull dimension 1 (i.e., not fields). If `R` is a
+field, its finite adèle ring is just defined to be the trivial ring.
+
+## References
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+
+## Tags
+finite adèle ring, dedekind domain
+-/
+
+noncomputable theory
+open function set is_dedekind_domain
+
+variables (R : Type) (K : Type) [comm_ring R] [is_domain R] [is_dedekind_domain R] [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+/-- The product of all `adic_completion_integers`, where `v` runs over the maximal ideals of `R`. -/
+def prod_completions_integers := (Π (v : height_one_spectrum R), v.adic_completion_integers K)
+local notation `R_hat` := prod_completions_integers
+
+instance : comm_ring (R_hat R K) := pi.comm_ring
+
+instance : topological_space (R_hat R K) := Pi.topological_space
+
+instance prod_completions_integers.inhabited : inhabited (prod_completions_integers R K) := ⟨0⟩
+
+/-- The product of all `adic_completion`, where `v` runs over the maximal ideals of `R`. -/
+def prod_completions := (Π (v : height_one_spectrum R), v.adic_completion K)
+
+local notation `K_hat` := prod_completions
+
+instance : comm_ring (K_hat R K) := pi.comm_ring
+instance : ring (K_hat R K) := infer_instance
+instance : topological_space (K_hat R K) := Pi.topological_space
+instance : topological_ring (K_hat R K) :=
+(infer_instance : topological_ring (Π (v : height_one_spectrum R), v.adic_completion K))
+
+instance prod_completions.inhabited : inhabited (prod_completions R K) := ⟨0⟩
+
+/-- The diagonal inclusion `r ↦ (r)_v` of `R` in `R_hat`. -/
+def diag_inj : R → (R_hat R K) := λ r v, inj_adic_completion_integers R K v r
```

Comment:
This is a `ring_hom`, right? Even better, you can maybe put an `R`-algebra structure on `R_hat` and define this as an `alg_hom`.

Reply:
Actually, now that the algebra instances are there, I don't think that this is needed (this is just the underlying map of `algebra_map R (R_hat R K)`).